// kepler.gl/dist/components/side-panel/layer-panel/column-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _localization = require("../../../localization");

var _ = require("../..");

var _fieldSelector = _interopRequireDefault(require("../../common/field-selector"));

var _visStateMerger = require("../../../reducers/vis-state-merger");

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  width: 75%;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  width: 25%;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  margin-bottom: 8px;\n  align-items: center;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var ColumnRow = _styledComponents["default"].div(_templateObject());

var ColumnName = _styledComponents["default"].div(_templateObject2());

var ColumnSelect = _styledComponents["default"].div(_templateObject3());

ColumnSelectorFactory.deps = [_fieldSelector["default"]];

function ColumnSelectorFactory(FieldSelector) {
  var ColumnSelector = function ColumnSelector(_ref) {
    var column = _ref.column,
        columns = _ref.columns,
        label = _ref.label,
        allFields = _ref.allFields,
        onSelect = _ref.onSelect,
        fieldPairs = _ref.fieldPairs;
    return /*#__PURE__*/_react["default"].createElement(ColumnRow, {
      className: "layer-config__column__selector"
    }, /*#__PURE__*/_react["default"].createElement(ColumnName, {
      className: "layer-config__column__name"
    }, /*#__PURE__*/_react["default"].createElement(_.PanelLabel, null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: "columns.".concat(label)
    })), !column.optional ? /*#__PURE__*/_react["default"].createElement(_.PanelLabel, null, "  *") : null), /*#__PURE__*/_react["default"].createElement(ColumnSelect, {
      className: "layer-config__column__select"
    }, /*#__PURE__*/_react["default"].createElement(FieldSelector, {
      suggested: fieldPairs,
      error: !(0, _visStateMerger.validateColumn)(column, columns, allFields),
      fields: allFields,
      value: column.value,
      erasable: Boolean(column.optional),
      onSelect: onSelect
    })));
  };

  return ColumnSelector;
}

var _default = ColumnSelectorFactory;
exports["default"] = _default;
