// kepler.gl/dist/components/side-panel/layer-panel/vis-config-switch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _infoHelper = _interopRequireDefault(require("../../common/info-helper"));

var _switch = _interopRequireDefault(require("../../common/switch"));

var _styledComponents2 = require("../../common/styled-components");

var _utils = require("../../../utils/utils");

var _localization = require("../../../localization");

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  justify-content: space-between;\n\n  .vis-config-switch__title {\n    display: flex;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var propTypes = {
  layer: _propTypes["default"].object.isRequired,
  property: _propTypes["default"].string.isRequired,
  onChange: _propTypes["default"].func.isRequired,
  label: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].bool, _propTypes["default"].func]),
  description: _propTypes["default"].string,
  disabled: _propTypes["default"].bool
};

var StyledVisConfigSwitch = _styledComponents["default"].div(_templateObject());

VisConfigSwitchFactory.deps = [_infoHelper["default"]];

function VisConfigSwitchFactory(InfoHelper) {
  var VisConfigSwitch = function VisConfigSwitch(_ref) {
    var _ref$layer = _ref.layer,
        id = _ref$layer.id,
        config = _ref$layer.config,
        property = _ref.property,
        _onChange2 = _ref.onChange,
        label = _ref.label,
        description = _ref.description,
        disabled = _ref.disabled;
    return /*#__PURE__*/_react["default"].createElement(_styledComponents2.SidePanelSection, {
      disabled: Boolean(disabled)
    }, /*#__PURE__*/_react["default"].createElement(StyledVisConfigSwitch, {
      className: "vis-config-switch"
    }, /*#__PURE__*/_react["default"].createElement("div", {
      className: "vis-config-switch__title"
    }, label ? /*#__PURE__*/_react["default"].createElement(_styledComponents2.PanelLabel, null, label && /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: label
    }) || (0, _utils.capitalizeFirstLetter)(property)) : null, description ? /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement(InfoHelper, {
      description: description,
      id: "".concat(id, "-").concat(property, "-description")
    })) : null), /*#__PURE__*/_react["default"].createElement("div", {
      className: "vis-config-switch__switch"
    }, /*#__PURE__*/_react["default"].createElement(_switch["default"], {
      checked: config.visConfig[property],
      id: "".concat(id, "-").concat(property, "-switch"),
      onChange: function onChange() {
        return _onChange2((0, _defineProperty2["default"])({}, property, !config.visConfig[property]));
      }
    }))));
  };

  VisConfigSwitch.propTypes = propTypes;
  return VisConfigSwitch;
}

var _default = VisConfigSwitchFactory;
exports["default"] = _default;
