// kepler.gl/dist/components/common/map-layer-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _checkbox = _interopRequireDefault(require("./checkbox"));

var _utils = require("../../utils/utils");

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 12px;\n\n  .map-layer-selector__item {\n    margin: 12px 0;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var propTypes = {
  // Required
  layers: _propTypes["default"].arrayOf(_propTypes["default"].object).isRequired,
  onMapToggleLayer: _propTypes["default"].func.isRequired
};

var MapLayerSelect = _styledComponents["default"].div(_templateObject());

var MapLayerSelector = function MapLayerSelector(_ref) {
  var layers = _ref.layers,
      onMapToggleLayer = _ref.onMapToggleLayer;
  return /*#__PURE__*/_react["default"].createElement(MapLayerSelect, {
    className: "map-layer-selector"
  }, layers.map(function (layer, index) {
    return /*#__PURE__*/_react["default"].createElement("div", {
      key: layer.id,
      className: "map-layer-selector__item"
    }, /*#__PURE__*/_react["default"].createElement(_checkbox["default"], {
      type: "radio",
      checked: layer.isVisible,
      id: "".concat(layer.id, "-toggle-").concat((0, _utils.generateHashId)(4)),
      label: layer.name,
      onChange: function onChange(e) {
        e.preventDefault();
        onMapToggleLayer(layer.id);
      }
    }));
  }));
};

MapLayerSelector.propTypes = propTypes;
var _default = MapLayerSelector;
exports["default"] = _default;
