// kepler.gl/dist/components/common/animation-control/playback-controls.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.AnimationWindowControl = exports.IconButton = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _classnames = _interopRequireDefault(require("classnames"));

var _localization = require("../../../localization");

var _styledComponents2 = require("../styled-components");

var _animationSpeedSlider = _interopRequireDefault(require("./animation-speed-slider"));

var _icons = require("../icons");

var _constants = require("../../../constants");

var _dataUtils = require("../../../utils/data-utils");

var _DEFAULT_ANIMATE_ITEM;

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  width: ", "px;\n  height: 32px;\n  color: ", ";\n  background-color: ", ";\n  border-radius: 4px;\n  margin-left: 7px;\n  border: 0;\n  padding: 0;\n\n  .__react_component_tooltip {\n    font-family: ", ";\n  }\n  svg {\n    margin: 0;\n  }\n  &.active {\n    background-color: ", ";\n  }\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  align-items: center;\n\n  .animation-control__speed-slider {\n    left: 0;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  position: relative;\n\n  &.disabled {\n    opacity: 0.4;\n    pointer-events: none;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var DELAY_SHOW = 500;
var DEFAULT_BUTTON_HEIGHT = '20px';

var StyledAnimationControls = _styledComponents["default"].div(_templateObject());

var StyledSpeedControl = _styledComponents["default"].div(_templateObject2());

var IconButton = (0, _styledComponents["default"])(_styledComponents2.Button)(_templateObject3(), function (props) {
  return props.collapsed ? 0 : 32;
}, function (props) {
  return props.theme.playbackButtonColor;
}, function (props) {
  return props.theme.playbackButtonBgColor;
}, function (props) {
  return props.theme.fontFamily;
}, function (props) {
  return props.theme.playbackButtonActBgColor;
});
exports.IconButton = IconButton;

function nop() {}

var DEFAULT_ICONS = {
  /* eslint-disable react/display-name */
  reset: function reset() {
    return /*#__PURE__*/_react["default"].createElement(_icons.Reset, {
      height: "18px"
    });
  },
  play: function play() {
    return /*#__PURE__*/_react["default"].createElement(_icons.Play, {
      height: "18px"
    });
  },
  pause: function pause() {
    return /*#__PURE__*/_react["default"].createElement(_icons.Pause, {
      height: "18px"
    });
  },

  /* eslint-enable react/display-name */
  speed: _icons.Rocket,
  animationFree: _icons.FreeWindow,
  animationIncremental: _icons.AnchorWindow
};
var DEFAULT_ANIMATE_ITEMS = (_DEFAULT_ANIMATE_ITEM = {}, (0, _defineProperty2["default"])(_DEFAULT_ANIMATE_ITEM, _constants.ANIMATION_WINDOW.free, {
  id: _constants.ANIMATION_WINDOW.free,
  icon: DEFAULT_ICONS.animationFree,
  tooltip: 'tooltip.animationByWindow'
}), (0, _defineProperty2["default"])(_DEFAULT_ANIMATE_ITEM, _constants.ANIMATION_WINDOW.incremental, {
  id: _constants.ANIMATION_WINDOW.incremental,
  icon: DEFAULT_ICONS.animationIncremental,
  tooltip: 'tooltip.animationByIncremental'
}), _DEFAULT_ANIMATE_ITEM);

var AnimationWindowControl = function AnimationWindowControl(_ref) {
  var _onClick = _ref.onClick,
      selected = _ref.selected,
      onHide = _ref.onHide,
      height = _ref.height,
      animationItems = _ref.animationItems,
      _ref$btnStyle = _ref.btnStyle,
      btnStyle = _ref$btnStyle === void 0 ? {} : _ref$btnStyle;
  return /*#__PURE__*/_react["default"].createElement("div", null, Object.values(animationItems).filter(function (item) {
    return item.id !== selected;
  }).map(function (item) {
    return /*#__PURE__*/_react["default"].createElement(IconButton, (0, _extends2["default"])({
      key: item.id,
      "data-tip": true,
      "data-for": "".concat(item.id, "-tooltip"),
      className: "playback-control-button",
      onClick: function onClick() {
        _onClick(item.id);

        onHide();
      }
    }, btnStyle), /*#__PURE__*/_react["default"].createElement(item.icon, {
      height: height
    }), item.tooltip ? /*#__PURE__*/_react["default"].createElement(_styledComponents2.Tooltip, {
      id: "".concat(item.id, "-tooltip"),
      effect: "solid",
      place: "top"
    }, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: item.tooltip
    })) : null);
  }));
};

exports.AnimationWindowControl = AnimationWindowControl;
PlaybackControlsFactory.deps = [_animationSpeedSlider["default"]];

function PlaybackControlsFactory(AnimationSpeedSlider) {
  // eslint-disable-next-line complexity
  var PlaybackControls = function PlaybackControls(_ref2) {
    var _ref2$isAnimatable = _ref2.isAnimatable,
        isAnimatable = _ref2$isAnimatable === void 0 ? true : _ref2$isAnimatable,
        isAnimating = _ref2.isAnimating,
        width = _ref2.width,
        speed = _ref2.speed,
        _ref2$animationWindow = _ref2.animationWindow,
        animationWindow = _ref2$animationWindow === void 0 ? _constants.ANIMATION_WINDOW.free : _ref2$animationWindow,
        setFilterAnimationWindow = _ref2.setFilterAnimationWindow,
        updateAnimationSpeed = _ref2.updateAnimationSpeed,
        _ref2$pauseAnimation = _ref2.pauseAnimation,
        pauseAnimation = _ref2$pauseAnimation === void 0 ? nop : _ref2$pauseAnimation,
        _ref2$resetAnimation = _ref2.resetAnimation,
        resetAnimation = _ref2$resetAnimation === void 0 ? nop : _ref2$resetAnimation,
        _ref2$startAnimation = _ref2.startAnimation,
        startAnimation = _ref2$startAnimation === void 0 ? nop : _ref2$startAnimation,
        _ref2$playbackIcons = _ref2.playbackIcons,
        playbackIcons = _ref2$playbackIcons === void 0 ? DEFAULT_ICONS : _ref2$playbackIcons,
        _ref2$animationItems = _ref2.animationItems,
        animationItems = _ref2$animationItems === void 0 ? DEFAULT_ANIMATE_ITEMS : _ref2$animationItems,
        _ref2$buttonStyle = _ref2.buttonStyle,
        buttonStyle = _ref2$buttonStyle === void 0 ? 'secondary' : _ref2$buttonStyle,
        _ref2$buttonHeight = _ref2.buttonHeight,
        buttonHeight = _ref2$buttonHeight === void 0 ? DEFAULT_BUTTON_HEIGHT : _ref2$buttonHeight;

    var _useState = (0, _react.useState)(false),
        _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
        showSpeedControl = _useState2[0],
        toggleSpeedControl = _useState2[1];

    var _useState3 = (0, _react.useState)(false),
        _useState4 = (0, _slicedToArray2["default"])(_useState3, 2),
        showAnimationWindowControl = _useState4[0],
        setShowAnimationWindowControl = _useState4[1];

    var toggleAnimationWindowControl = (0, _react.useCallback)(function () {
      setShowAnimationWindowControl(!showAnimationWindowControl);
    }, [showAnimationWindowControl, setShowAnimationWindowControl]);
    var btnStyle = buttonStyle ? (0, _defineProperty2["default"])({}, buttonStyle, true) : {};
    return /*#__PURE__*/_react["default"].createElement(StyledAnimationControls, {
      className: (0, _classnames["default"])('playback-controls', {
        disabled: !isAnimatable
      }),
      style: {
        width: "".concat(width, "px")
      }
    }, setFilterAnimationWindow ? /*#__PURE__*/_react["default"].createElement(IconButton, (0, _extends2["default"])({
      "data-tip": true,
      "data-for": "animate-window",
      className: (0, _classnames["default"])('playback-control-button', {
        active: showAnimationWindowControl
      }),
      onClick: toggleAnimationWindowControl
    }, btnStyle), function () {
      if (animationItems[animationWindow]) {
        return /*#__PURE__*/_react["default"].createElement(animationItems[animationWindow].icon, {
          height: buttonHeight
        });
      }

      return null;
    }(), animationItems[animationWindow] && animationItems[animationWindow].tooltip ? /*#__PURE__*/_react["default"].createElement(_styledComponents2.Tooltip, {
      id: "animate-window",
      place: "top",
      delayShow: 500,
      effect: "solid"
    }, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: animationItems[animationWindow].tooltip
    })) : null) : null, showAnimationWindowControl ? /*#__PURE__*/_react["default"].createElement(AnimationWindowControl, {
      onClick: setFilterAnimationWindow,
      selected: animationWindow,
      onHide: toggleAnimationWindowControl,
      height: buttonHeight,
      btnStyle: btnStyle,
      animationItems: animationItems
    }) : null, showAnimationWindowControl || !updateAnimationSpeed ? null : /*#__PURE__*/_react["default"].createElement(StyledSpeedControl, {
      onClick: function onClick() {
        toggleSpeedControl(!showSpeedControl);
      }
    }, /*#__PURE__*/_react["default"].createElement(IconButton, (0, _extends2["default"])({
      "data-tip": true,
      "data-for": "animate-speed",
      className: "playback-control-button",
      onClick: nop
    }, btnStyle), /*#__PURE__*/_react["default"].createElement(playbackIcons.speed, {
      height: buttonHeight
    }), /*#__PURE__*/_react["default"].createElement(_styledComponents2.Tooltip, {
      id: "animate-speed",
      place: "top",
      delayShow: DELAY_SHOW,
      effect: "solid"
    }, /*#__PURE__*/_react["default"].createElement("span", null, (0, _dataUtils.preciseRound)(speed, 1), "x"))), showSpeedControl ? /*#__PURE__*/_react["default"].createElement(AnimationSpeedSlider, {
      onHide: function onHide() {
        toggleSpeedControl(false);
      },
      updateAnimationSpeed: updateAnimationSpeed,
      speed: speed
    }) : null), showAnimationWindowControl ? null : /*#__PURE__*/_react["default"].createElement(IconButton, (0, _extends2["default"])({
      "data-tip": true,
      "data-for": "animate-reset",
      className: "playback-control-button",
      onClick: resetAnimation
    }, btnStyle), /*#__PURE__*/_react["default"].createElement(playbackIcons.reset, {
      height: buttonHeight
    }), /*#__PURE__*/_react["default"].createElement(_styledComponents2.Tooltip, {
      id: "animate-reset",
      place: "top",
      delayShow: DELAY_SHOW,
      effect: "solid"
    }, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: "tooltip.reset"
    }))), showAnimationWindowControl ? null : /*#__PURE__*/_react["default"].createElement(IconButton, (0, _extends2["default"])({
      "data-tip": true,
      "data-for": "animate-play-pause",
      className: (0, _classnames["default"])('playback-control-button', {
        active: isAnimating
      }),
      onClick: isAnimating ? pauseAnimation : startAnimation,
      hide: showSpeedControl
    }, btnStyle), isAnimating ? /*#__PURE__*/_react["default"].createElement(playbackIcons.pause, {
      height: buttonHeight
    }) : /*#__PURE__*/_react["default"].createElement(playbackIcons.play, {
      height: buttonHeight
    }), /*#__PURE__*/_react["default"].createElement(_styledComponents2.Tooltip, {
      id: "animate-play-pause",
      place: "top",
      delayShow: DELAY_SHOW,
      effect: "solid"
    }, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: isAnimating ? 'tooltip.pause' : 'tooltip.play'
    }))));
  };

  return PlaybackControls;
}

var _default = PlaybackControlsFactory;
exports["default"] = _default;
