// react-map-gl/dist/es6/components/fullscreen-control.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { document } from '../utils/globals';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { useEffect, useState, useMemo } from 'react';
import mapboxgl from '../utils/mapboxgl';
import useMapControl, { mapControlDefaultProps, mapControlPropTypes } from './use-map-control';
const propTypes = Object.assign({}, mapControlPropTypes, {
  className: PropTypes.string,
  style: PropTypes.object,
  container: PropTypes.object,
  label: PropTypes.string
});
const defaultProps = Object.assign({}, mapControlDefaultProps, {
  className: '',
  container: null,
  label: 'Toggle fullscreen'
});

function FullscreenControl(props) {
  const {
    context,
    containerRef
  } = useMapControl(props);
  const [isFullscreen, setIsFullscreen] = useState(false);
  const [showButton, setShowButton] = useState(false);
  const [mapboxFullscreenControl, createMapboxFullscreenControl] = useState(null);
  useEffect(() => {
    const container = props.container || context.container;
    const control = new mapboxgl.FullscreenControl({
      container
    });
    createMapboxFullscreenControl(control);
    setShowButton(control._checkFullscreenSupport());

    const onFullscreenChange = () => {
      const nextState = !control._fullscreen;
      control._fullscreen = nextState;
      setIsFullscreen(nextState);
    };

    document.addEventListener(control._fullscreenchange, onFullscreenChange);
    return () => {
      document.removeEventListener(control._fullscreenchange, onFullscreenChange);
    };
  }, []);

  const onClickFullscreen = () => {
    if (mapboxFullscreenControl) {
      mapboxFullscreenControl._onClickFullscreen();
    }
  };

  const style = useMemo(() => _objectSpread({
    position: 'absolute'
  }, props.style), [props.style]);

  if (!showButton) {
    return null;
  }

  const {
    className,
    label
  } = props;
  const type = isFullscreen ? 'shrink' : 'fullscreen';
  return React.createElement("div", {
    style: style,
    className: className
  }, React.createElement("div", {
    className: "mapboxgl-ctrl mapboxgl-ctrl-group",
    ref: containerRef
  }, React.createElement("button", {
    key: type,
    className: "mapboxgl-ctrl-icon mapboxgl-ctrl-".concat(type),
    type: "button",
    title: label,
    onClick: onClickFullscreen
  }, React.createElement("span", {
    className: "mapboxgl-ctrl-icon",
    "aria-hidden": "true"
  }))));
}

FullscreenControl.propTypes = propTypes;
FullscreenControl.defaultProps = defaultProps;
export default React.memo(FullscreenControl);
